cpp/src/arrow/compute/kernels/compare.cc

```diff
@@ -17,250 +17,296 @@
   });
-
   return Status::OK();
 }
 
 template <typename ArrowType, CompareOperator Op>
-class CompareFunctionImpl final : public CompareFunction {
+class CompareKernel final : public BinaryKernel {
+ public:
   using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
   using ScalarType = typename TypeTraits<ArrowType>::ScalarType;
 
- public:
-  explicit CompareFunctionImpl(FunctionContext* ctx) : ctx_(ctx) {}
-
-  Status Compare(const ArrayData& array, const Scalar& scalar, ArrayData* output) const {
-    // Caller must cast
-    DCHECK(array.type->Equals(scalar.type));
-    // Output must be a boolean array
-    DCHECK(output->type->Equals(boolean()));
-    // Output must be of same length
-    DCHECK_EQ(output->length, array.length);
-
-    // Scalar is null, all comparisons are null.
-    if (!scalar.is_valid) {
-      return detail::SetAllNulls(ctx_, array, output);
-    }
+  std::shared_ptr<DataType> out_type() const override { return boolean(); }
 
-    // Copy null_bitmap
-    RETURN_NOT_OK(detail::PropagateNulls(ctx_, array, output));
+  Status Call(FunctionContext* ctx, const Datum& left, const Datum& right,
+              Datum* out_datum) override {
+    auto out = out_datum->array();
 
-    uint8_t* bitmap_result = output->buffers[1]->mutable_data();
-    return CompareArrayScalar<ArrowType, Op>(
-        array, static_cast<const ScalarType&>(scalar), bitmap_result);
-  }
+    auto left_array = AsArray(left);
+    auto left_scalar = AsScalar(left);
+
+    auto right_array = AsArray(right);
+    auto right_scalar = AsScalar(right);
 
-  Status Compare(const Scalar& scalar, const ArrayData& array, ArrayData* output) const {
-    // Caller must cast
-    DCHECK(array.type->Equals(scalar.type));
-    // Output must be a boolean array
-    DCHECK(output->type->Equals(boolean()));
-    // Output must be of same length
-    DCHECK_EQ(output->length, array.length);
-
-    // Scalar is null, all comparisons are null.
-    if (!scalar.is_valid) {
-      return detail::SetAllNulls(ctx_, array, output);
+    if (left_array && right_array) {
+      RETURN_NOT_OK(AssignNulls(ctx, *left_array, *right_array, out.get()));
+      return Compare<Op>(MakeRange(*left_array), MakeRange(*right_array), out.get());
     }
 
-    // Copy null_bitmap
-    RETURN_NOT_OK(detail::PropagateNulls(ctx_, array, output));
+    if (left_array && right_scalar) {
+      RETURN_NOT_OK(AssignNulls(ctx, *right_scalar, *left_array, out.get()));
+      return Compare<Op>(MakeRange(*left_array), MakeRange(*right_scalar), out.get());
+    }
 
-    uint8_t* bitmap_result = output->buffers[1]->mutable_data();
-    return CompareScalarArray<ArrowType, Op>(static_cast<const ScalarType&>(scalar),
-                                             array, bitmap_result);
-  }
+    if (left_scalar && right_array) {
+      RETURN_NOT_OK(AssignNulls(ctx, *left_scalar, *right_array, out.get()));
+      return Compare<Op>(MakeRange(*left_scalar), MakeRange(*right_array), out.get());
```

Comment:
Will "scalar CMP array" generate separate code from "array CMP scalar"? Just wondering whether we can limit the number of different specializations.

Reply:
I think this could be accomplished generally by providing a mapping from op -> inverse op, so that `5 > [1, 4, 5, 6, 9]` could be replaced by `[1, 4, 5, 6, 9] < 5`. @fsaintjacques